src/libstd/io/buffered.rs

```diff
@@ -818,6 +899,261 @@ impl<W> fmt::Display for IntoInnerError<W> {
+    /// means that, if any newlines are present in the data, the data up to
+    /// and including the buffer containing the last newline is sent directly
+    /// to the inner writer, and the data after it is buffered. Returns the
+    /// number of bytes written.
+    ///
+    /// This function operates on a "best effort basis"; in keeping with the
+    /// convention of `Write::write`, it makes at most one attempt to write
+    /// new data to the underlying writer.
+    ///
+    /// Because this function attempts to send completed lines to the underlying
+    /// writer, it will also flush the existing buffer if it contains any
+    /// newlines.
+    ///
+    /// Because sorting through an array of `IoSlice` can be a bit convoluted,
+    /// This method differs from write in the following ways:
+    ///
+    /// - It attempts to write all the buffers up to and including the one
+    ///   containing the last newline. This means that it may attempt to
+    ///   write a partial line.
+    /// - If the write only reports partial success, it does not attempt to
+    ///   find the precise location of the written bytes and buffer the rest.
+    ///
+    /// If the underlying vector doesn't support vectored writing, we instead
+    /// simply write the first non-empty buffer with `write`. This way, we
+    /// get the benefits of more granular partial-line handling without losing
+    /// anything in efficiency
+    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
+        // If there's no specialized behavior for write_vectored, just use
+        // write. This has the benefit of more granular partial-line handling.
+        if !self.is_write_vectored() {
+            return match bufs.iter().find(|buf| !buf.is_empty()) {
+                Some(buf) => self.write(buf),
+                None => Ok(0),
+            };
+        }
+
+        // Find the buffer containing the last newline
+        let last_newline_buf_idx = bufs
+            .iter()
+            .enumerate()
+            .rev()
+            .find_map(|(i, buf)| memchr::memchr(b'\n', buf).map(|_| i));
+
+        // If there are no new newlines (that is, if this write is less than
+        // one line), just do a regular buffered write
+        let last_newline_buf_idx = match last_newline_buf_idx {
+            // No newlines; just do a normal buffered write
+            None => {
+                // Check for prior partial line writes that need to be retried.
+                // Only retry if the buffer contains a completed line, to
+                // avoid flushing partial lines.
+                if let Some(b'\n') = self.buffered().last().copied() {
+                    self.buffer.flush_buf()?;
+                }
+                return self.buffer.write_vectored(bufs);
+            }
+            Some(i) => i,
+        };
+
+        // Flush existing content to prepare for our write
+        self.buffer.flush_buf()?;
```

Comment:
So there are a couple problems with doing it here; one is specific to `write_vectored`, and the other applies to all uses of this technique:

- Because our input is (equivalent to) `&[&[u8]]`, the only way to prepend our pre-buffered data would be allocating a new vec<&[u8]>, which we want to avoid for hopefully obvious reasons. This problem doesn't apply to `write` or `write_all`, because our input is `&[u8]`, which means we can construct a `[&[u8]; 2]` to send to `write_vectored`.
- The key difference between `flush_buf` and `write` is that `write` makes at-most 1 attempt, whereas `flush_buf` has to loop until the buffer is empty or an error occurs. This isn't insurmountable, but it means that `flush_buf` has to be `vector aware`, so that it can make looping calls to `write_vectored` but STOP when the writes transition to new data.

Reply:
Another argument in favor of this is my proposal to replace `Vec<u8>` with `VecDeque<u8>` in `BufWriter`. This is a circular buffering that would eliminate the need for internal data copies in `BufWriter`, and substantially simplify the implementation of `flush_buf`. One of the downsides is that it means that the buffer can be "split", when the circular buffer wraps around the edge; vectored writes would mitigate this issue.

Reply:
An alternative to stitching via vectored writes would be only bypassing the buffer if it is empty and otherwise stitching things together in the buffer when it fits and then flushing the buffer up to the last newline in it.

Reply:
Same problem. If we buffer first, we *must* report that we've consumed data to the caller, which means we're back in the position of having to ignore i/o errors if they occur after buffering the data. This behavior was the cause of a lot of complexity of the old `LineWriter` implementation, and before that it was the cause of a lot of bugs (#32085, #37807), so removing it was a major goal for this refactor.

Reply:
I didn't mean really buffering the data, just using BufWriter's buffer as scratch-space to finish a partial line and then flushing that, this would again result in a single actual write. Of course this kind of mending would only be possible if the line length is shorter than the buffer size.

If we don't want to allocate a new [&IoVec] to merge the buffer into vectored writes then the alternative is completing the line inside the buffer and falling back to a single-buffer write if the buffer is non-empty.

Basically we would only be doing vectored writes if the buffer is empty or when the lines are so huge that they don't fit into the buffer anyway. This also makes sense since vectored writes do not make sense if the caller isn't giving us complete lines, it's relying on LineWriter to do the assembly. Vectored writes only make sense if the caller sends a whole pile of lines and relies on LineWriter to slice things in a way that the lines are written atomically.

Actually, thinking about it, LineWriter shouldn't try to get out as many lines as possible. If pipes are involved (not sure how ptys behave) it should only try to aggregate lines up to `PIPE_BUF` bytes to avoid tearing.

Reply:
- Regarding using the buffer as scratch space. Suppose we've buffered `"AAAA"`, and suppose that a write comes in for `"BBBB\nCCC"`. If we append `"BBBB\n"` to the buffer, we *must* report that we've consumed 5 bytes to the caller, by (at some point) returning `Ok(5)`. This means that, if the call to `inner.write("AAAABBBB\n") returns an error, we either have to silently drop that error, or remove the `"BBBB\n"` from the buffer. Additionally, we have to make sure that we retry `Ok(2)` (because we only wrote pre-buffered data) but *don't* retry `Ok(6)` (because we should aim to make at-most 1 attempt to write new data).
- Regarding vectored writes: I agree, it's difficult to reconcile vectored writes with the behavior of LineBuffer. A reasonable behavior would be to disable them entirely, and allow it to simply forward to write. However, I'm assuming that users who want to combine vectored writes with `LineWriter` "know what they're doing", and so the behavior I've implemented tries to find the best balance of directly forwarding writes, without a lot of complex processing in the middle, with doing what you'd expect a `LineWriter` to do. It's worth noting that in the (presumably) common case that the buffers with newlines are terminated by them, the logic as-implemented exactly matches `write`.
- `LineWriter` doesn't have anything to do with unix pipelines; it's completely destination agnostic. It can be used to write to pipes, or to files, or to TCP sockets, or anything else. The only thing it's supposed to do is be a variant `BufWriter` that sends data to the wrapper writer when a newline comes through. The rest of the behavior about trying not to write partial lines, and trying only to write at newline boundaries, are design decisions I've made that I think represent best practice for a writer of this type.
- Related to what you're talking about is *specifically* what stdout should be doing. This pull request is preliminary work required for a future Pull Request for #60673, which aims to enable line-buffered vs block-buffered mode for stdout, which will (presumably) be set by default based on the tty status of stdout (and something similar on windows).

Reply:
> This means that, if the call to inner.write("AAAABBBB\n") returns an error, we either have to silently drop that error, or remove the "BBBB\n" from the buffer. 

I was thinking of the latter, which should be a fairly cheap operation, only the length needs to be changed on the buffer.

> Related to what you're talking about is specifically what stdout should be doing.

Not necessarily. If you have a linewriter wrapping a pipe fd then you only get atomic write guarantees up to `PIPE_BUF`. So if the expected behavior of LineWriter is to "write one line at a time" then aggregating multiple lines into a single write could violate the expectation of getting your lines written atomically. If too many lines get aggregated this can lead to unintentional interleaving (which looks a lot like tearing) in the pipe.
So in some circumstances it may make sense to deaggregate lines into multiple smaller writes (by doing short writes, no need to do multiple write calls internally) to provide per-line atomicity. Especially when we already have (partial) lines buffered from a previous write.

This isn't specific to stdout, it's useful for any kind of concurrent writing of line-based output. If you use a logger which has thread-local `LineWriter<&File/Pipe/whatever>` instances then avoiding line interleaving requires atomicity. Note that there is no locking.

> LineWriter doesn't have anything to do with unix pipelines; it's completely destination agnostic. It can be used to write to pipes, or to files, or to TCP sockets, or anything else. 

Indeed, it's not possible to get any atomicity guarantees from sockets under concurrent access. But regular files in append mode and pipes do offer some atomicity and it would be nice if linewriter could preserve those properties.

Reply:
My thinking would be that that use case should be handled by another `Write` adapter, or even a dedicated type. Certainly `LineBuffer` shouldn't change its behavior in the presence of environment variable or things like that. The `write` contract doesn't claim to provide "atomicity", because it doesn't prescribe what device is being wrapped. It just describes a composable method by which many write calls may be made, and a consistent state be retained even in the presence of partial writes or failures. If your use cases involves shared access to a single writable handle (atypical, because `Write` takes methods by `&mut self`), that would call for an adapter that caps write lengths (perhaps returning an error for too-large writes).

Reply:
> My thinking would be that that use case should be handled by another Write adapter,

But if you want line-based atomicity you need to search your input buffers for the right place to split the writes, so it would mostly be replicating the string scanning that LineWriter already does anyway.

> atypical, because Write takes methods by &mut self)

`impl<'_> Write for &'_ File`, thus having concurrent writers for any file descriptor is part of the standard library.

> In the former case, the user plainly doesn't want atomic line writes

Not necessarily. They could be expecting linewriter to provide that atomicity. If you do

```rust
writer.write(b"foo")?;
writer.write(b"bar")?;
writer.write(b"\n")?;
```

Then it will do it in a single syscall as you'd expect it. It's atomic on a good day.


> The device wrote less than a full line.

Not all types of file descriptors encounter short writes. At least linux akes an effort¹ to make regular file writes an all-or-nothing affair and for pipes there's the above-mentioned `PIPE_BUF` guarantee. Perhaps the pipe case can be covered by making an attempt to search for a newline within `capacity` bytes. I.e. if the BufWriter capacity is set to 8KiB and a caller passes a 32KiB of data then LineWriter it will do a short write for the first N lines that fit into 8KiB and only do a longer unbuffered write if it can't find one in that window.

Then the stdout implementation could probe if the file descriptor is a pipe and set the buffer size accordingly to `PIPE_BUF`, which will then be taken as a hint to deaggregate lines if needed.
An explicit parameter would work too (0 = no short writes for deaggregation).
Of course this is just a nice-to-have. But in my experience interleaved lines from multithreaded logging is a nuisance that can be avoided exactly this way.


¹ I say effort because there have been bugs and some network filesystems where that isn't the case. But local filesystems uphold this.